elastic/kibana
# Add user_action app for gathering telemetry on user actions

## In this PR

Fixes https://github.com/elastic/kibana/issues/31330

- Adds an OSS app called `user_action`, which publishes a single API endpoint: `POST /user_action/<app>/<action_type>`
- Updates the Rollup Job Wizard to hit this endpoint when the tries to create a rollup job
- Updates the Rollup Job Wizard's usage collector to publish this metric

In the collector, this takes the form of:

```json
{
  "user_actions": {
    "create_rollup_job": 4
  }
}
```

## To-do

- [x] Add API integration tests
- [ ] ~~Per feedback from Jon B., throttle requests to one every 5 seconds~~ Unlike sending an API request per keystroke as someone types, these actions will be dispatched when users click buttons and tabs, which is unlikely to exceed a request every few seconds.

## Testing

1. Create a rollup job (Management > Rollup Jobs) with no "Delay", and a "Frequency" of "Every minute".
2. Visit `http://localhost:5601/api/stats?extended=true` and look for these stats to reflect the saved objects you've created.

To see all of the generated user action saved objects, go to Dev Tools and run this:

```
GET _search
{
  "query": {
    "match": {
      "type": "user-action"
    }
  }
}
```


# Dev Docs
## Add user_action app for gathering telemetry on user actions
Added a new API endpoint `POST /user_action/<app>/<action_type>` which is used to add UI telemetry to track user actions. This API that is not intended for external use and will likely change in the next versions.
